feat(xo-server): implement rolling pool reboot

fbeauchamp opened the pull request:
### Description

_Short explanation of this PR (feel free to re-use commit message)_

### Checklist

- Commit
  - Title follows [commit conventions](https://bit.ly/commit-conventions)
  - Reference the relevant issue (`Fixes #007`, `See xoa-support#42`, `See https://...`)
  - If bug fix, add `Introduced by`
- Changelog
  - If visible by XOA users, add changelog entry
  - Update "Packages to release" in `CHANGELOG.unreleased.md`
- PR
  - If UI changes, add screenshots
  - If not finished or not tested, open as _Draft_

fbeauchamp committed:
feat(xo-server): implement rolling pull reboot

MathieuRA commented on packages/xo-server/src/xapi/mixins/patching.mjs:
```diff
@@ -651,4 +637,22 @@ export default {
   },
+
+  @decorateWith(deferrable)
+  async rollingPoolUpdate($defer, { xsCredentials } = {}) {
```
`hasMissingPatchesByHost` should be calculated only on `rollingPoolUpdate`, otherwise hosts will not restart in some cases.

MathieuRA commented on packages/xo-server/src/xapi/mixins/patching.mjs:
```diff
@@ -493,9 +493,7 @@ export default {
-  async rollingPoolUpdate($defer, { xsCredentials } = {}) {
-    const isXcp = _isXcp(this.pool.$master)
-
+  async roolingPullReboot($defer, { xsCredentials, beforeEvacuateAny, beforeEvacuateHost } = {}){
```
I think, `rollingPollReboot` shoud be on `mixin/pool.mjs` and not on `mixin/patching.mjs`

MathieuRA reviewed (commented):
Since we are very close to the release, what do you think if for the `rolling Pool Reboot` we just copy/paste the necessary code, and think about how to mutualise the code right after the release?

fbeauchamp commented:
> Since we are very close to the release, what do you think if for the `rolling Pool Reboot` we just copy/paste the necessary code, and think about how to mutualise the code right after the release?

if possible, I would like to try the refactoring, the rolling pool update method is already complex enough . If we're not confident on the code quality , we can still fallback to an additionnal flag on wednesday

julien-f commented:
I would love the logic to be mutualized with `host_smartReboot` as well.

fbeauchamp commented:
> I would love the logic to be mutualized with `host_smartReboot` as well.


how do you see things  ? since the code logic seems quite different , I think I need more time to think this out. Each Vm will be handled exactly twice in smartreboot ( suspend, then resume), but an unlucky VM can be migrated a lot of times in case of rolling pull reboot . Also the error cases are different (allowed operation Vs local storage)

I think both feature should be blended : in case of rolling pool update/reboot , we could fall back to suspend  for example

MathieuRA commented on packages/xo-server/src/xo-mixins/pool.mjs:
```diff
@@ -134,4 +134,12 @@ export default class Pools {
     )
   }
+
+  async rollingPullReboot(pool) {
```
`rollingPullReboot` -> `rollingPoolReboot`

MathieuRA commented on packages/xo-server/src/xapi/mixins/pool.mjs:
```diff
@@ -22,4 +32,145 @@ export default {
   },
+  
+  @decorateWith(deferrable)
+  async rollingPullReboot($defer, { xsCredentials, beforeEvacuateAny, beforeRebootHost, ignoreHost } = {}){
```
Same.

MathieuRA commented on packages/xo-server/src/xapi/mixins/patching.mjs:
```diff
@@ -496,159 +492,30 @@ export default {
-            error = err
-          }
+      },
+      beforeRebootHost: async host=>{
```
```suggestion
      beforeRebootHost: async host => {
```

MathieuRA commented on packages/xo-server/src/xapi/mixins/patching.mjs:
```diff
@@ -496,159 +492,30 @@ export default {
-          }
+      },
+      beforeRebootHost: async host=>{
+        if(isXcp){
```
```suggestion
        if (isXcp) {
```

MathieuRA commented on packages/xo-server/src/xapi/mixins/patching.mjs:
```diff
@@ -496,159 +492,30 @@ export default {
-        if (residentVms.includes(vmId)) {
-          continue
+    await this.rollingPullReboot({ xsCredentials ,  
+      beforeEvacuateAny: async ()=>{
```
```suggestion
      beforeEvacuateAny: async () => {
```

MathieuRA commented on packages/xo-server/src/api/pool.mjs:
```diff
@@ -200,6 +200,31 @@ rollingUpdate.resolve = {
+rollingReboot.params = {
+  bypassBackupCheck: {
+    optional: true,
+    type: 'boolean',
```
Set the default value directly here.

MathieuRA reviewed (changes requested):
Not tested.

MathieuRA commented on packages/xo-server/src/api/pool.mjs:
```diff
@@ -200,6 +200,32 @@ rollingUpdate.resolve = {
+  bypassBackupCheck: {
+    optional: true,
+    type: 'boolean',
+    default: false
```
If `default` is defined, `optional` is not needed.
```suggestion
    default: false
    type: 'boolean',
```

MathieuRA commented on packages/xo-server/src/xapi/mixins/patching.mjs:
```diff
@@ -496,159 +492,30 @@ export default {
+      beforeRebootHost: async host => {
+        if (isXcp) {
+          
+          await this.installPatches({ hosts: [host] })
```
```suggestion
          log.debug(`Install patches on host ${hostId}`)
          await this.installPatches({ hosts: [host] })
```

MathieuRA commented:
Don't forget to rebase on master to get the "Rolling pool update" bug fix.

```js
// host.$resident_VMs is outdated and returns resident VMs before the host.evacuate.
// this.getField is used in order not to get cached data.
const residentVmRefs = await this.getField('host', host.$ref, 'resident_VMs')

fbeauchamp commented:
test done : rolling reboot et rolling pull update works

julien-f commented on packages/xo-server/src/xapi/mixins/pool.mjs:
```diff
@@ -21,9 +30,153 @@ const methods = {
     })
   },
+
+  async rollingPoolReboot($defer, { xsCredentials, beforeEvacuateAny, beforeRebootHost, ignoreHost } = {}) {
```
`xsCredentials` is unused.

julien-f committed:
remove unused var

julien-f committed:
beforeEvacuateAny → beforeEvacuateVms

This makes it more clear that it is called only once and not one time per host or VM.

julien-f committed:
re-add comment/log
